CASSANDRA-14654 Reduce heap pressure during compactions

clohfink opened the pull request.

krummas commented on src/java/org/apache/cassandra/io/sstable/SSTableRewriter.java:
```diff
@@ -131,16 +132,20 @@ public RowIndexEntry append(UnfilteredRowIterator partition)
         // we do this before appending to ensure we can resetAndTruncate() safely if the append fails
         DecoratedKey key = partition.partitionKey();
-        maybeReopenEarly(key);
+        if (preemptiveOpenInterval != Long.MAX_VALUE)
```
this is checked in `maybeReopenEarly` - no need to check here as well

krummas commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -238,6 +238,7 @@
     public int hints_flush_period_in_ms = 10000;
     public int max_hints_file_size_in_mb = 128;
     public ParameterizedClass hints_compression;
+    public boolean invalidate_cache_on_compaction = true;
```
this should be volatile for hotprops

krummas replied:
The biggest performance impact of setting this to `false` is probably not the fact that we don't invalidate the cache anymore, but that we don't warm the cache up for the new sstables - maybe we should rename the property? (warm_keycache_on_compaction or migrate_keycache_on_compaction?) Or at least add a comment about it? And it should probably say "keycache" instead of "cache"

krummas reviewed (commented):
we should probably add a few tests for this, EncodingStats.collect looks like a candidate for example

I pushed a commit here: https://github.com/krummas/cassandra/commits/chris/encodingstats to reduce the number of EncodingStats allocations a bit more, have not run any benchmarks, but it avoids creating an EncodingStats for every SSTableIdentityIterator we create

clohfink committed:
Reduce heap pressure during compactions
Patch by Chris Lohfink; Reviewed by Dinesh Joshi and Benedict for CASSANDRA-14654

clohfink committed:
dtests
